Usage documentation updates for the readme (#51)
airtasker/spot: README.md

[![License](https://img.shields.io/npm/l/@airtasker/spot.svg)](https://github.com/airtasker/spot/blob/master/package.json)

<!-- toc -->

- [Usage](#usage)
- [Commands](#commands)
  <!-- tocstop -->

# Usage

To get started and set up an API declaration in the current directory, run:

```
npx @airtasker/spot init
```

You can then run a generator with:

```
npx @airtasker/spot generate --api api.ts
```

## `@api`

Define an API. This is required and must only be defined once:

```TypeScript
import { api } from "@airtasker/spot";

> @api({

Reviewer: Now that `@api()` is used separately from `@endpoint()`, the syntax looks a bit unnecessarily complicated.

Something to address separately: maybe we actually just want to expose a conventional function `declareApi()`, so you do the following:
```typescript
declareApi({
  name: "My API",
  description: "My really cool API"
});
```

Author: Adding test-interactions will force us to potentially introduce some new syntax - which may or may not work well with the current syntax. It makes sense at that point that we evaluate the syntax overall to ensure both consistency and intuitiveness.

  name: "My API",
  description: "My really cool API"
})
class MyAPI {}
```

| Field         | Description                           |
| ------------- | ------------------------------------- |

> | `name`        | (**required**) Name of the API        |

Reviewer: Why are these two required? Just thinking in terms of reducing friction for users who don't need to generate documentation. Description isn't required by OpenAPI3 either: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#infoObject.

This is fine for now, but I'd like to address this in the future. It may be more of a philosophical question, but I think Spot should only require the absolute minimum, which is the structure of the API without any human-readable names or descriptions.

Author: Description makes sense to be optional, but I think name should be required. Without a name for the service/api, we can't realistically convert to other forms of contracts e.g. OpenAPI and Pact both require a service name/title (unless of course we give it a default name but could potentially be unsafe when working with a contract broker holding contracts for multiple services)

Reviewer: Makes sense.

| `description` | (**required**) Description of the API |

## `@endpoint`

Define a HTTP endpoint for the API. An endpoint describes a particular HTTP action on a URL path:

```TypeScript
import { endpoint, genericError, header, pathParam, queryParam, request, response, specificError } from "@airtasker/spot";

class MyUserEndpoints {

>   @endpoint({

Reviewer: It may be useful to add a comment before each of these endpoint definitions, paraphrasing what the declaration does.

```suggestion
  // GET /users expects a mandatory `search_term` query parameter and returns a list of users.
  @endpoint({
```

Note #1: I wrote "mandatory", but I wonder if `search_term` should be `Optional<string>` to reflect real endpoints more accurately? It's rare for a search query parameter to be mandatory.

Note #2: Should we instead return a `UserListResponse`? Returning a list of responses doesn't feel very idiomatic.

    method: "GET",
    path: "/users",
    description: "Retrieve all users",
    tags: ["Users"]
  })
  getUsers(@queryParam({ description: "Search term" }) search_term: string): UserResponse[] {
    return response();
  }

  @endpoint({
    method: "GET",
    path: "/users/:id",
    description: "Get user by id",
    tags: ["Users"]
  })
  getUser(@pathParam({ description: "Unique user identifier" }) id: string): UserResponse {
    return response();
  }

  @endpoint({
    method: "POST",
    path: "/users",
    description: "Create a user",
    tags: ["Users"]
  })
  @specificError<ApiError>({
    name: "unauthorized",
    statusCode: 401
  })
  createUser(
    @request req: CreateUserRequest,
    @header({
      name: "Authorization",
      description: "This is the authorization token"
    })
    authToken: Optional<string>
  ): CreateUserResponse {
    return response();
  }
}

interface UserResponse {
  firstName: string;
  lastName: string;
}

interface CreateUserRequest {
  firstName: string;
  lastName: string;
}

interface CreateUserResponse {
  success: boolean;
}

> interface ApiError {

Reviewer: Should this be renamed to `ApiErrorResponse`?

  message: string;
}
```

| Field                | Default            | Description                                               |
| -------------------- | ------------------ | --------------------------------------------------------- |
| `method`             |                    | (**required**) [HTTP method](#suppported-http-methods)    |
| `path`               |                    | (**required**) URL path                                   |
| `description`        | `""`               | Description of the endpoint                               |
| `requestContentType` | `application/json` | Content type of the request body                          |
| `successStatusCode`  | `200`              | HTTP status code for a successful response                |

> | `tags`               |                    | Array of tags used for endpoint grouping in documentation |

Reviewer: Nit: I personally find the term `List` less implementation-specific than `Array` (to me, "array" implies that indexes are important). Thoughts?

Author: Makes perfect sense

### `@request`

> Define a request body for requests that require one. This is commonly used for `POST` and `PUT` requests:

Reviewer: Maybe also add a note that it isn't valid for `GET` requests.

```TypeScript
class MyUserEndpoints {
  //...
  @endpoint({
    method: "POST",
    path: "/users"
  })
  createUser(
    @request req: CreateUserRequest
  ): UserResponse {
    return response();
  }

  @endpoint({
    method: "PUT",
    path: "/users/:id"
  })
  updateUser(
    @pathParam({ description: "User unique identifier" }) id: string,
    @request req: UpdateUserRequest
  ): UserResponse {
    return response();
  }
  //...
}

interface CreateUserRequest {
  firstName: string;
  lastName: string;
}

interface UpdateUserRequest {
  lastName: string;
}
```

> ### `@header`

Reviewer: Would it be possible to add examples for each of these too, or mention that the examples above cover them?

Define a request header. `@header` can be used multiple times to define multiple headers.

| Field         | Description                       |
| ------------- | --------------------------------- |
| `name`        | (**required**) Name of the header |
| `description` | Description of the header         |

### `@pathParam`

> Define path parameters that appear in the `@endpoint.path`. The parameter must be defined in the `@endpoint.path`. A `@pathParam` should be defined for each parameter defined in `@endpoint.path`.

Reviewer: It may not be super obvious that `@endpoint.path` refers to `@endpoint({ path: ... })` so I'd suggest the following:
```suggestion
Define path parameters that appear in the `path` provided in `@endpoint()`. For example if the path is `/users/:userId`, the endpoint method must define a matching argument with `@pathParam() userId: string`.

Note: the name of the argument must match the name of the path parameter.
```

| Field         | Description                       |
| ------------- | --------------------------------- |
| `description` | Description of the path parameter |

### `@queryParam`

> Define query parameters. `@queryParam` can be used multiple times to define multiple query parameters.

Reviewer: ```suggestion
Define query parameters. `@queryParam` can be used multiple times to define multiple query parameters.

Note: the name of the argument must match the name of the query parameter.
```

| Field         | Description                        |
| ------------- | ---------------------------------- |
| `description` | Description of the query parameter |

### `@specificError<T>`

> Define a known error for the endpoint. `@specificError` can be used multiple times to define multiple errors.

Reviewer: ```suggestion
Define a known error for the endpoint. `@specificError` can be used multiple times to define multiple errors.

`T` must be replaced with the response type when the error occurs, for example `@specificError<UnauthenticatedErrorResponse>`.
```

| Field        | Description                                   |
| ------------ | --------------------------------------------- |
| `name`       | (**required**) Name of the error              |
| `statusCode` | (**required**) HTTP status code for the error |

### `@genericError<T>`

> Define a default error for the endpoint. This can only be used once for an `@endpoint`.

Reviewer: ```suggestion
Define a default error for the endpoint. This can only be used once for an `@endpoint`.

`T` must be replaced with the response type when the error occurs, for example `@specificError<UnauthenticatedErrorResponse>`.
```

## Matcher Types

| Type          | Description                                                                                             | Example                                                |
| ------------- | ------------------------------------------------------------------------------------------------------- | ------------------------------------------------------ |
| `string`      | A string value                                                                                          |
| `number`      | A number value                                                                                          |

> | `Int32`       | A 32-bit integer                                                                                        |

Reviewer: The inconsistency between lower and uppercase types makes me wonder if we could rename our types to be lowercase instead... Alternatively, we could support the uppercase aliases, which already exist: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

| `Int64`       | A 64-bit integer                                                                                        |
| `Float`       | A 32-bit floating point number                                                                          |
| `Double`      | A 64-bit floating point number                                                                          |
| `boolean`     | A boolean value                                                                                         |

> | `Date`        | [ISO-8601](https://www.iso.org/iso-8601-date-and-time-format.html) string representation of a date      |

Reviewer: These could use examples (as well as each of the above).

    path: "/users"
  })
  createUser(
    @request req: CreateUserRequest
  ): UserResponse {
    return response();
  }

  @endpoint({
    method: "PUT",
    path: "/users/:id"
  })
  updateUser(
    @pathParam({ description: "User unique identifier" }) id: string,
    @request req: UpdateUserRequest
  ): UserResponse {
    return response();
  }
  //...
}

interface CreateUserRequest {
  firstName: string;
  lastName: string;
}

interface UpdateUserRequest {
  lastName: string;
}
```

### `@header`

Define a request header. `@header` can be used multiple times to define multiple headers:

```TypeScript
  //...
  @endpoint({
    method: "POST",
    path: "/users",
    description: "Create a user"
  })
  createUser(
    @request req: CreateUserRequest,
    @header({
      name: "Authorization",
      description: "This is the authorization token"
    })
    authToken: Optional<string>
  ): CreateUserResponse {
    return response();
  }
  //...
```

| Field         | Description                       |
| ------------- | --------------------------------- |
| `name`        | (**required**) Name of the header |
| `description` | Description of the header         |

### `@pathParam`

> Define path parameters that appear in the `path` provided in `@endpoint()`. For example if the path is `/users/:userId`, the endpoint method must define a matching argument with `@pathParam() userId: string`:

Reviewer: ```suggestion
Define path parameters that appear in the `path` provided in `@endpoint()`. For example if the path is `/users/:id`, the endpoint method must define a matching argument with `@pathParam() id: string`:
```

(to be consistent with the example below)

```TypeScript
  //...
  @endpoint({
    method: "GET",
    path: "/users/:id",
    description: "Get user by id"
  })
  getUser(@pathParam({ description: "Unique user identifier" }) id: string): UserResponse {
    return response();
  }
  //...
```

**Note**: the name of the argument must match the name of the path parameter.

| Field         | Description                       |
| ------------- | --------------------------------- |
| `description` | Description of the path parameter |

### `@queryParam`

Define query parameters. `@queryParam` can be used multiple times to define multiple query parameters:

```TypeScript
  //...
  @endpoint({
    method: "GET",
    path: "/users",
    description: "Retrieve all users"
  })
  getUsers(@queryParam({ description: "Search term" }) search_term: Optional<string>): UserResponse[] {
    return response();
  }
  //...
```

> **Note**: the name of the argument must match the name of the query parameter.

Reviewer 2: Is this true? I don't think there is any argument that needs to match the name of the query param. Maybe it is for path param

| Field         | Description                        |
| ------------- | ---------------------------------- |
| `description` | Description of the query parameter |

### `@specificError<T>`

> Define a known error for the endpoint. `@specificError` can be used multiple times to define multiple errors. `T` must be replaced with the response type when the error occurs, for example `@specificError<UnauthenticatedErrorResponse>`:

Reviewer: ```suggestion
Define a known error for the endpoint. `@specificError` can be used multiple times to define multiple errors. `T` must be replaced with the response type when the error occurs, for example `@specificError<UnauthorizedErrorResponse>`:
```

```TypeScript
  //...
  @endpoint({
    method: "POST",
    path: "/users",
    description: "Create a user"
  })

>   @specificError<UnauthenticatedErrorResponse>({

Reviewer: ```suggestion
  @specificError<UnauthorizedErrorResponse>({
```

```
USAGE
$ spot init

OPTIONS
-h, --help show CLI help

EXAMPLE
$ spot init
Generated the following files:

- api.ts
- tsconfig.json
- package.json
```

_See code: [cli/src/commands/init.js](https://github.com/airtasker/spot/blob/master/cli/src/commands/init.ts)_

<!-- commandsstop -->

> ```

Reviewer 2: there is an extra code line here
